broker-core/src/main/java/org/apache/qpid/server/virtualhost/VirtualHostConnectionLimiter.java

```diff
@@ -0,0 +1,294 @@
+        {
+            _limiter.set(newLimiter(_connectionLimitProviders));
+        }
+    }
+
+    private ConnectionLimiter newLimiter(final Map<ConnectionLimitProvider, ConnectionLimiter> cache)
+    {
+        ConnectionLimiter limiter = ConnectionLimiter.noLimits();
+
+        LOGGER.debug("Updating virtual host connection limiters");
+        for (final VirtualHostConnectionLimitProvider<?> provider :
+                _virtualHost.getChildren(VirtualHostConnectionLimitProvider.class))
+        {
+            if (provider.getState() == State.ACTIVE)
+            {
+                limiter = limiter.append(
+                        cache.computeIfAbsent(provider, ConnectionLimitProvider::getConnectionLimiter));
+            }
+            else if (provider.getState() == State.ERRORED)
+            {
+                limiter = ConnectionLimiter.blockEveryone();
+            }
+        }
+
+        LOGGER.debug("Updating broker connection limiters");
+        for (final BrokerConnectionLimitProvider<?> provider :
+                _broker.getChildren(BrokerConnectionLimitProvider.class))
+        {
+            if (provider.getState() == State.ACTIVE)
+            {
+                limiter = limiter.append(
+                        cache.computeIfAbsent(provider, ConnectionLimitProvider::getConnectionLimiter));
+            }
+            else if (provider.getState() == State.ERRORED)
+            {
+                limiter = ConnectionLimiter.blockEveryone();
+            }
+        }
+        return ConnectionLimiter.cacheConnectionRegistration(limiter);
+    }
+
+    private abstract static class AbstractChangeListener extends AbstractConfigurationChangeListener
+    {
+        private final VirtualHostConnectionLimiter _limiter;
+
+        AbstractChangeListener(VirtualHostConnectionLimiter limiter)
+        {
+            super();
+            _limiter = Objects.requireNonNull(limiter);
+        }
+
+        void addProvider(ConfiguredObject<?> provider)
+        {
+            provider.addChangeListener(new ProviderChangeListener(_limiter));
+            _limiter.update();
+        }
+
+        void removeProvider(ConfiguredObject<?> provider)
+        {
+            provider.removeChangeListener(new ProviderChangeListener(_limiter));
+            _limiter.update(provider);
+        }
+
+        void updateProvider(ConfiguredObject<?> provider)
+        {
+            _limiter.update(provider);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return _limiter.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj)
+        {
+            if (obj instanceof AbstractChangeListener)
+            {
+                return _limiter == ((AbstractChangeListener) obj)._limiter;
+            }
+            return false;
+        }
+    }
+
+    private static final class VirtualHostChangeListener extends AbstractChangeListener
```

Comment:
The VirtualHostChangeListener code is a duplicate of BrokerChangeListener.
It seems it can be parametrised using category class, and child category class. Those can be passed into constructor